Always load Rascal modules and Java classes of `std` from the current `rascal` (v2)

sungshik opened the pull request:
### Overview

This PR implements the following approach for loading Rascal modules of `rascal`:

https://github.com/usethesource/rascal/blob/d328b34e6e125146369ef8f5ed94421ce6e7c34d/src/org/rascalmpl/library/util/PathConfig.java#L667-L691

See also [this comment](https://github.com/usethesource/rascal/pull/2112#issuecomment-2618186651) for additional background/motivation.

Java classes of `rascal` are always loaded from the "current" `rascal` (as per `resolveCurrentRascalRuntimeJar()`).

Thus, after this PR:
  - ✔️ Rascal modules and Java classes of `rascal` inside `std` must be loaded from the same version. This partly fixes usethesource/rascal-language-servers#538.

  - ⚠️ Rascal modules and Java classes of `rascal` outside `std` may be loaded from different versions. However, currently, there are no Rascal modules of `rascal` outside `std` that rely on Java classes. Moreover, after the merge, there are a few Rascal modules of `rascal-core` that rely on Java classes, but the impact seems negligible in the context of this PR.
    
  - ⚠️ REPLs created in VS Code always load the version of `std` of the deployed `rascal.jar` (part of the VS Code extension), which is put on the class path of the REPL's JVM. Even when `rascal` itself is open in the workspace, changes made to `std` will *not* be loaded in the REPL. To test such changes, a standalone Rascal shell needs to be created (but it doesn't have debugging support).

### Examples (VS Code)

#### Create REPL in `rascal` (`typepal` open)

  - [ ] No compiler and TypePal
  - [ ] Compiler and TypePal from deployed `rascal.jar`
  - [x] Compiler and TypePal from open folders in workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |project://rascal/src/org/rascalmpl/compiler| at |file:///c:/Users/sung-/Desktop/rascal-which-version/projects/rascal/src/org/rascalmpl/compiler|
    |project://typepal/src| at |file:///c:/Users/sung-/Desktop/rascal-which-version/projects/typepal/src|
    |jar+file:///C:/Users/sung-/Desktop/rascal-language-servers3/rascal-vscode-extension/assets/jars/rascal-lsp.jar!/|
```

#### Create REPL in `rascal` (`typepal` closed)

  - [x] No compiler and TypePal
  - [ ] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |jar+file:///C:/Users/sung-/Desktop/rascal-language-servers3/rascal-vscode-extension/assets/jars/rascal-lsp.jar!/|
```

#### Create REPL in `rascal-lsp` (`rascal` open, `typepal` open)

  - [ ] No compiler and TypePal
  - [ ] Compiler and TypePal from deployed jars
  - [x] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |project://rascal/src/org/rascalmpl/compiler| at |file:///c:/Users/sung-/Desktop/rascal-which-version/projects/rascal/src/org/rascalmpl/compiler|
    |project://typepal/src| at |file:///c:/Users/sung-/Desktop/rascal-which-version/projects/typepal/src|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `rascal-lsp` (`rascal` open, `typepal` closed)

  - [ ] No compiler and TypePal
  - [x] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/compiler|
    |jar+file:///C:/Users/sung-/.m2/repository/org/rascalmpl/typepal/0.14.0/typepal-0.14.0.jar!/src|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `rascal-lsp` (`rascal` closed, `typepal` open)

  - [ ] No compiler and TypePal
  - [x] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/compiler|
    |jar+file:///C:/Users/sung-/.m2/repository/org/rascalmpl/typepal/0.14.0/typepal-0.14.0.jar!/src|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `rascal-lsp` (`rascal` closed, `typepal` closed)

  - [ ] No compiler and TypePal
  - [x] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/compiler|
    |jar+file:///C:/Users/sung-/.m2/repository/org/rascalmpl/typepal/0.14.0/typepal-0.14.0.jar!/src|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `typepal` (`rascal` open)

  - [x] No compiler, TypePal from workspace
  - [ ] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/typepal/src|
    |jar+file:///C:/Users/sung-/Desktop/rascal-language-servers3/rascal-vscode-extension/assets/jars/rascal-lsp.jar!/|
```

#### Create REPL in `typepal` (`rascal` closed)

  - [x] No compiler, TypePal from workspace
  - [ ] Compiler and TypePal from deployed jars
  - [ ] Compiler and TypePal from workspace

```
Module paths:
    |std:///| at |jar+file:///C:/Users/sung-/Desktop/rascal3/rascal/target/rascal-0.40.8-SNAPSHOT.jar!/org/rascalmpl/library/|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/typepal/src|
    |jar+file:///C:/Users/sung-/Desktop/rascal-language-servers3/rascal-vscode-extension/assets/jars/rascal-lsp.jar!/|
```

### Examples (`RascalShell`)

#### Create REPL in `rascal`

```
Module paths:
    |std:///| at |file:///C:/Users/sung-/Desktop/rascal3/rascal/target/classes/org/rascalmpl/library/|
```

#### Create REPL in `rascal-lsp` (with TypePal jar on the class path)

```
Module paths:
    |std:///| at |file:///C:/Users/sung-/Desktop/rascal3/rascal/target/classes/org/rascalmpl/library/|
    |file:///C:/Users/sung-/Desktop/rascal3/rascal/target/classes/org/rascalmpl/compiler|
    |jar+file:///C:/Users/sung-/.m2/repository/org/rascalmpl/typepal/0.14.0/typepal-0.14.0.jar!/src|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `rascal-lsp` (without TypePal jar on the class path)

```
Module paths:
    |std:///| at |file:///C:/Users/sung-/Desktop/rascal3/rascal/target/classes/org/rascalmpl/library/|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/rascal-language-servers/rascal-lsp/src/main/rascal|
```

#### Create REPL in `typepal`

```
Module paths:
    |std:///| at |file:///C:/Users/sung-/Desktop/rascal3/rascal/target/classes/org/rascalmpl/library/|
    |file:///C:/Users/sung-/Desktop/rascal-which-version/projects/typepal/src|
```

sungshik commented:
This PR is intended to replace #2103. The main change relative to that PR is, per @jurgenvinju's suggestion, non-`std` source folders of `rascal` will never be loaded in any REPL, except when the main project of the REPL is `rascal` itself.

Instead of giving this special treatment only to `std`, the implementation of this PR is a bit more general and currently gives the same special treatment to (a temporary placeholder for) `typepal`. The idea is that, beside `std`, there may be other source folders in `rascal` that are useful to "normal" users.

DavyLandman commented:
I think I like this approach, but I don't think typepal should be treated in the future like std lib. Specifically, you want to be able to work on typepal itself while inside VS Code. No need to treat it in a special way like we do the std lib.

sungshik commented:
Ah yes, good point. I'll do that.

jurgenvinju commented:
Great work!

> Rascal modules of rascal inside special source folders are always loaded from the current rascal (as per resolveCurrentRascalRuntimeJar).

This one I don't understand. I'm expecting we will read Rascal modules from `std:///` for as long as we have the interpreter in play, and by extension we'd use `typepal:///`, etc. to refer to libraries inside the rascal.jar. This `|std:///|` stuff is necessary because `resolveCurrentRascalRuntimeJar + path` produces a `jar+file` URI that is specific for the local machine, breaking source debugging and references in the LSP to library functions and globals and modules.

sungshik commented:
I've implemented Davy's suggestion and updated the issue description accordingly. I think Jurgen's main remaining point is to introduce a `typepal` scheme. The same thought also crossed my mind yesterday, but after talking with Davy about it briefly, the idea was abandoned. Let's first discuss this a bit more before I start implementing it 😎.

sungshik commented on src/org/rascalmpl/library/util/PathConfig.java:
```diff
@@ -635,15 +637,22 @@ else if (rascalProject != null) {
-            }
+        if (!projectName.equals("rascal")) { // Don't add `rascal` again
+            for (String srcName : manifest.getSourceRoots(manifestRoot)) {
+                var srcFolder = URIUtil.getChildLocation(manifestRoot, srcName);
+                
+                if (!reg.exists(srcFolder) || !reg.isDirectory(srcFolder)) {
+                    messages.append(Messages.error("Source folder " + srcFolder + " does not exist.", getRascalMfLocation(rascalProject)));
+                }
 
-            srcsWriter.append(srcFolder);
+                srcsWriter.append(srcFolder);
+            }
```
(Exactly the same as before, but indented because of the outer conditional)

DavyLandman commented:
Let me explain my stance on this a bit more verbosely (although I've tried to keep it short).

1. I dislike `|std:///|` and `|lib://typepal|` as it doesn't tell you which version/where it came from. For that reason I much prefer the `mvn://...` scheme, and also the `jar+file://` or `|project://` ones (even though project can be a bit confusing). As in, `std:///` and `lib://typepal` are a bit magic for end users to figure out where files came from.
2. If there is a problem with typecheckers that requires us to introduce `typepal:///` (and others), I think that it would also break for external libraries (such as salix & rascal-lsp). 
3. It looks a bit like we're going from `lib://typepal` to `typepal://`, which defeats the purpose of getting rid of the lib scheme a bit?

> The LSP breaks because the .tpl files for typepal will have different hard locations for every declaration; those locations will be either mvn:/// or std:/// or typepal:/// but never file:///C:/Users/sung-/

To me this looks like more a typepal/rascal-core feature that should be able to relatize it's locations (as it already does, since it stores the locations in a logical form, and maps it to a physical at the last moment). But not something that we should try and solve by introducing a new kind of special location.

jurgenvinju commented:
> I dislike |std:///| and |lib://typepal| as it doesn't tell you which version/where it came from. For that reason I much prefer the mvn://... scheme, and also the jar+file:// or |project:// ones (even though project can be a bit confusing). As in, std:/// and lib://typepal are a bit magic for end users to figure out where files came from.

I don't like it that much either, but we need a **installation-location-independent URI** for the libraries inside rascal.jar. Since they end up in the extension folder and not in the `mvn` repo, the `std:///` location  was invented. The same location is also workable for rascal.jar in its independent mode ("java -jar rascal.jar"), etc. 

Features like the debugger and browsing through the standard libraries with M3 and .tpl summaries depend on this.

jurgenvinju commented:
It is possible to turn std:/// and typepal:/// and stuff into **logical** URI resolvers. That way they become _transparant_ and can be queried for their actual location. Since authorities for logical resolvers are also keys, we could have _one_ scheme for the rascal jar and then an authority per nested library:
* |boot://std/| for the standard library
* |boot://typepal| for the typepal sources
* |boot://compiler| for the compiled sources

When we `resolve` these logical locations you will get `jar+file:///path/to/rascal.jar!/src/org/rascalmpl/library`, etc. So we can explain to the user which exact versions are used at the moment.

I like "boot" since there is a bootstrap dependency on these libraries for making rascal work. It is very similar to the bootstrap bytecode/dll classpath of the JVM.

jurgenvinju commented:
By the way if we change std:// to boot://std we have to do a big bang release of everything, i just realized. So better stick with std:// for the next few releases.

DavyLandman commented:
I feel like we're moving towards a solution that has the same problems as the lib schema. While there is a case for the rascal.jar, I think we don't need the special case (as it hides a current limitation of rascal-core). But I do think a few issues are getting mixed here.

Different issues that we're trying to squash (using `lib://clair` as an example):

1. Inside of VS Code, `lib://clair` is not resolvable. As in, which version of clair does it point to? especially since the VFS resolver is part of the JVM that runs the typechecker & outliner, it doesn't even have clair on it's class/source path. So, our fix was (in this PR series): let's replace that with `mvn://.../clair-0.1.0.jar` (or `project://clair`).
3. When starting a REPL, you have to very carefully read all the output to know where clair actually came from. Our fix (in this PR series): let's get rid of `lib://clair` and the whole `resolved lib://clair...` message and use only the path config data.
4. While clair was typechecked, the sources were located in `file:///.../path/on/ci/clair/src`. That path is not accurate outside of the build environment. Historically we fixed this by rewriting those path (in the package phase) to the `lib://clair` scheme. However in VS Code, this doesn't work, as `lib://clair` was not something that VS Code could navigate to (see point 1), meaning we cannot navigate to clair or any other third party libraries. 

I think point 3 is not an issue we should try solve in rascal itself with logical source locations. I think the work @PaulKlint has done to store logical locations (`rascal:///util/Monitor` instead of `file:///..../util/Monitor.rsc`) can be adapted to generate a proper source paths for any tpl based on the context of where the tpl is found. 

<details>
<summary>sketch on how I imagine this would work in rascal-core</summary>

1. we use `rascal://util//Monitor` like we currently do
2. the mapping to physical locations doesn't contain `file:///..../util/Monitor.rsc` locations, but `relative:///util/Monitor.rsc` locations. (optionally this is only done during packaging)
3. When we ask for use defs/errors, we translate the relative locations based on the current location of the tpls/sources (most likely a source path)

I'll leave the details of this to @PaulKlint but based on our discussions, this should be possible.
</details>


How does this relate to `boot://rascal` or `std:///` or `boot://typepal`? I think for 1 and 2 they confuse the end user of where stuff is coming from and are not needed. For 3 it's a workaround that only works for things prepackaged with the extension. 


I'm sorry I couldn't phrase this as a reply to previous messages. But I'll now try to link it to some specific quotes by @jurgenvinju

> but we need a installation-location-independent URI for the libraries inside rascal.jar
> Features like the debugger and browsing through the standard libraries with M3 and .tpl summaries depend on this.

Why specifically? Can you give a concrete example of why it needs to be installation-location-independent? I know we currently depend on the `std:///` but I think we do not have to. 

> It is (since we have PathConfig.getCurrentRascalRuntime()) possible to turn std:/// and typepal:/// and stuff into logical URI resolvers.

This only holds as long as everyone has to use the exact same version of rascal (and typepal) as the rascal extension is using. I know that's a current requirement, but I would like to move to a place where we don't need this. I think if we don't depend on `std:///` being the same for all REPLs it would make life easier down the road. (Same is true for typechecking code with a slightly older/newer version of rascal than the typechecker is running under).

sungshik commented:
After some in-person talks to discuss how to move forward, here's a summary of the plan.

### The merge

  - There are three kinds of projects:
      - Rascal (`rascal`)
      - Rascal developer projects (`rascal-core`, `typepal`, `rascal-lsp`)
      - Rascal user projects (any other project)
  - During the merge:
      - `rascal-core` moves into `rascal`.
      - `typepal` and `rascal-lsp` stay separate projects.
  - After the merge:
      - `rascal` depends on `typepal`, cyclically. To avoid Maven issues, instead of listing `typepal` as a dependency in `rascal`'s POM, the required `typepal` version is copied into the jar when `rascal` is packaged. This copy of `typepal` is never loaded in REPLs of Rascal user projects, though (i.e., Rascal user projects may depend on a version of `typepal` of their own choosing).
      - `rascal-lsp` no longer depends on `rascal-core` and `typepal`; only on `rascal`.
      - There are three source folders in `rascal` that may be loaded in the REPL:
          - `std:///`
          - `project://rascal/.../compiler` (never loaded by REPLs of Rascal user projects)
          - `project://rascal/.../typepal-copy` (never loaded by REPLs of Rascal user projects)

### Rules to load source folders of `rascal`/`typepal` in REPLs after the merge

  - If a REPL is created in `rascal`:
      - If using VS Code, and if `project://typepal/` does exists, then load:
          - `std:///` from deployed jar
          - `project://rascal/.../compiler` from open folder
          - `project://typepal/` from open folder
      - If using VS Code, and if  `project://typepal/` doesn't exist, then load only `std:///` from deployed jar.
      - If using Rascal Shell, then load only `std:///` from open folder.
  - If a REPL is created in `rascal-lsp`:
      - If `project://rascal/` and `project://typepal/` do both exist, then load:
          - `std:///` from deployed jar
          - `project://rascal/.../compiler` from open folder
          - `project://typepal/` from open folder
      - If `project://rascal/` or `project://typepal/` doesn't exist, then load:
          - `std:///` from deployed jar
          - `project://rascal/compiler` from deployed jar
          - `project://rascal/.../typepal-copy` from deployed jar
  - If a REPL is created in `typepal` or in a Rascal user project, then load only `std:///` from deployed jar.

### Consequences for REPLs in VS Code

  - `std` is always loaded from the deployed `rascal.jar` (and guaranteed to be consistent with the deployed `rascal-lsp.jar`).
  - The combination of `compiler` and `typepal` is guaranteed to be consistent.
  - The combination of `std` and `compiler`/`typepal` isn't guaranteed to be consistent (because the former can come from the deployed `rascal.jar`, while the latter may come from the workspace).

sungshik commented on src/org/rascalmpl/shell/ShellEvaluatorFactory.java:
```diff
@@ -28,34 +28,30 @@
         if (rootFolder != null) {
             configureProjectEvaluator(evaluator, rootFolder);
+        } else {
+            evaluator.addRascalSearchPathContributor(StandardLibraryContributor.getInstance());
```
Moved this addition of the standard library to this else branch (instead of always adding it), because `configureProjectEvaluator` adds it, too, and the search path would include both `|std:///|` and `std`.

sungshik commented on src/org/rascalmpl/shell/ShellEvaluatorFactory.java:
```diff
@@ -28,34 +28,30 @@
+        return createEvaluator(input, stdout, stderr, monitor, rootFolder);
+    }
+
+    private static Evaluator createEvaluator(InputStream input, OutputStream stdout, OutputStream stderr, IRascalMonitor monitor, ISourceLocation rootFolder) {
```
Created this small private method as I found myself making the same changes twice

sungshik commented on src/org/rascalmpl/shell/ShellEvaluatorFactory.java:
```diff
@@ -74,9 +70,6 @@ public static void configureProjectEvaluator(Evaluator evaluator, ISourceLocatio
             evaluator.addRascalSearchPath((ISourceLocation) path); 
         }
 
-        for (IValue path : pcfg.getLibs()) {
-            evaluator.addRascalSearchPath((ISourceLocation) path);
-        }
```
The source path of `pcfg` already contains the source folders of libs (added [here](https://github.com/usethesource/rascal/blob/80ad94ee2e22c4aca9b577d9284041f48a7fe6e3/src/org/rascalmpl/library/util/PathConfig.java#L547) and [there](https://github.com/usethesource/rascal/blob/80ad94ee2e22c4aca9b577d9284041f48a7fe6e3/src/org/rascalmpl/library/util/PathConfig.java#L576) when the path config is computed), so their addition is already covered by the previous loop

sungshik reviewed (commented):
A few comments to clarify

jurgenvinju commented:
Ok good progress. 

Only let's not copy source files into the rascal project. That will be a great source of confusion when the copy is edited. Rather let's edit them from their original source project. 

For the Rascal project we can add `project://typepal/src` to the PathConfig srcrs. If it exists. If it doesn't exist we should not add it and also not add the compiler sources, and put a warning about this into the messages list.

We could also check the typepal dependency version against the version of the source and add warnings to the messages field if they don't lign up.

sungshik commented:
Ok, great, thanks Jurgen! I've implemented your suggestion under the assumption that we'll start shipping a `typepal.jar` inside the VSIX, which is put on the class path of a REPL process. This requires a few small changes in the corresponding rascal-language-servers branch; I'll make those when this PR is merged into your branch. Unless there are other comments, I think we can go ahead and merge.

jurgenvinju commented:
Note that we are already shipping typepal.jar with the vsix

jurgenvinju reviewed (approved):
Let's merge this; we might run into some things along the way but I think it's better to bite the bullet. 

@DavyLandman this is what you proposed right? To move forward knowing that the integration on VScode level could be a bit unstable from here on? I think we can merge this and the parent branch into main, and then move on to the branches on the maven-plugin for the next step in supporting the mvn:// scheme and the error reporting for pom.xml and rascal.mf

DavyLandman commented:
yes, agreed, it's too hard to avoid all kind of breakage here.
